src/new.rs

```diff
@@ -520,18 +528,28 @@ pub fn hello_world_test() {{
   |> should.equal("Hello, from {name}!")
 }}
 "#,
-                name = self.options.name
+                name = self.options.name.clone().unwrap()
             ),
         )
     }
 }
 
-pub fn create(options: NewOptions, version: &'static str) -> Result<()> {
-    validate_name(&options.name)?;
-    validate_root_folder(&options.name)?;
-    let creator = Creator::new(options, version);
+pub fn create(mut options: NewOptions, version: &'static str) -> Result<()> {
+    if options.name.is_none() {
+        options.name = get_foldername(&options.project_root);
```

Comment:
Because we are storing name to be used as an optional it needs to be cloned and unwrapped at each time it is used. I don't mind the cleaning so much because this code only runs once, but I wish to avoid the use of unwrap because it is unsafe and could potentially crash. While we know that the name does exist at this point there is no way of verifying that using the type system when using unwrap￼.

Rather than determining the name to use and then storing it as an option let's store it just as a string. This way we know that it always exists when we need it in the other functions.￼￼

Reply:
> Rather than determining the name to use and then storing it as an option let's store it just as a string. This way we know that it always exists when we need it in the other functions.￼￼
Hi @lpil,
I'm not sure what you mean by that, the point of this PR is making the name of the project optional so it could be inferred from the filesystem if needed, also decoupling the name of the project from the base folder making possible to have folder names like `my-project`, the easy way I found was using a default value ("") for `name` but you didn't agree with that, would you mind to explain yourself a bit more about what you are suggesting now, please? :)

Reply:
Yes, that's right. We want the name to be optional, falling back to the name of the project root directory if the name is not given.

Here's some pseudocode

```rust
let name = if let Some(name) = config.name {
  name.to_string()
} else {
  name_from
};
validate_name(&name)?;
let creator = Creator::new(options.clone(), name, version);
creator.run()?;
```

It is the same approach as using an empty string but it is clearer because the optional type indicates that this value may not be present, rather than having one string value be handled differently.